### PR TITLE
chore: upgrade test containers from buster to bookworm

### PR DESCRIPTION
**Pull request by jusiskin**
### What was the problem/requirement? (What/Why)

The developer scripting for test automation uses Docker to setup Linux containers used for cross-user test automation. The `Dockerfiles` build on the `python:3.7-buster` base container image. This container image is now end-of-life and a new image was last pushed a year ago (https://hub.docker.com/_/python/tags?name=3.9-buster)

### What was the solution? (How)

Modified the `Dockerfile`s to use `python:3.7-bookworm`.
    1. This required adding `-y` to `apt-get install ...` commands in the `Dockerfile`s since this was a behavior change from buster to bookworm
    2. The `test_subprocess.py` test module needed to be modified due to a behavior change between the base images. 
Processes killed with signals are put in a "zombie" state if their parent is still running but hasn't checked the exit code of the process (see https://en.wikipedia.org/wiki/Zombie_process)

While doing this, I had attempted to optimize the developer loop by caching the hatch environment in the built container image. I was unable to find a solution however, since hatch installs the project files and dependencies when creating the environment and there is no way to tell it to just install the dependencies. This means that anytime a source or test file are changed, the Docker image layer containing the hatch environment needs to be rebuilt.

I've left the changes to the `Dockerfile` restructuring in the event that hatch provides a feature for this or we switch developer tooling (e.g. to `uv` has [support for this optimization](https://docs.astral.sh/uv/guides/integration/docker/#intermediate-layers)). I've tested making a change to the `src` and `test` subdirectories, this only requires rebuilding the layer that creates the hatch environment.

### What is the impact of this change?

1.  The development workflow for cross-user Linux tests uses actively maintained container base images
2.  We have made incremental progress towards optimizing the development container workflow to reduce the time for a developer test cycle

### How was this change tested?

See [DEVELOPMENT.md](https://github.com/OpenJobDescription/openjd-model-for-python/blob/mainline/DEVELOPMENT.md#testing) for information on running tests.

- Ran the unit tests which succeded
- Ran the cross-user Linux tests using the modified scripting documented in `DEVELOPER.md` which succeeded

### Was this change documented?

- Code comments were added were deemed fit
- Test container README was updated
- Updated `DEVELOPMENT.md` since the `--build` argument was removed (it is always implied now)

### Is this a breaking change?

*  There were no breaking changes made to consumers of the library.
*  A breaking change to the development workflow was made (`--build` argument from `scripts/run_sudo_tests.sh` was removed and is now implied)

### Does this change impact security?

This change does not impact the threat model of the package 

This change improves the development-time security posture since developers will be using an actively maintained base container image.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

**Review by crowecawcaw — DISMISSED** [review]

File: scripts/run_sudo_tests.sh [hunk 1]
```diff
@@ -11,22 +11,19 @@ then
...
             ;;
-        --build)
-            DO_BUILD="True"
-            ;;
         --ldap)
             echo "Using the LDAP client container image for testing."
             USE_LDAP="True"
             ;;
         --build-only)
+            DO_BUILD="True"
```
Comment (ddneilson) [inline scripts/run_sudo_tests.sh]:
```suggestion
```

Looks like you got rid of all uses of `DO_BUILD`

File: testing_containers/ldap_sudo_environment/Dockerfile [hunk 2]
```diff
@@ -1,24 +1,24 @@
 # Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 
-FROM python:3.9-buster
+FROM python:3.9-bookworm
+ARG PIP_INDEX_URL
```
Comment (ddneilson) [inline testing_containers/ldap_sudo_environment/Dockerfile:4]: Q: Have you tested this build without having `PIP_INDEX_URL` defined in your environment?

File: testing_containers/ldap_sudo_environment/Dockerfile [hunk 3]
```diff
@@ -1,24 +1,24 @@
...
 
-COPY changePassword.ldif /config
-COPY addUsersGroups.ldif /config
-COPY addUsersToSharedGroup.ldif /config
-COPY run_tests.sh  /config
-COPY start_ldap.sh /config
-COPY test_setup.sh /config
-
-WORKDIR /code
+COPY testing_containers/ldap_sudo_environment/changePassword.ldif /config
```
Comment (ddneilson) [inline testing_containers/ldap_sudo_environment/Dockerfile:18]: Question: Why the switch to fuller relative paths in the source filenames? This makes the Dockerfile less movable.

**Review by ddneilson — CHANGES_REQUESTED** [review]

File: testing_containers/ldap_sudo_environment/Dockerfile (hunk 3 above)
Comment (jusiskin) [reply testing_containers/ldap_sudo_environment/Dockerfile:18]: This was in order to copy the source tree into the container which removes file-system ownership (UID/GID) and permission coupling between the host machine and the container.

File: testing_containers/ldap_sudo_environment/Dockerfile (hunk 2 above)
Comment (jusiskin) [reply testing_containers/ldap_sudo_environment/Dockerfile:4]: Just tested and the workflow succeeds end-to-end.

File: scripts/run_sudo_tests.sh (hunk 1 above)
Comment (jusiskin) [reply scripts/run_sudo_tests.sh]: Ah good catch. Removed this last occurrence.

File: testing_containers/ldap_sudo_environment/Dockerfile (hunk 2 above)
Comment (ddneilson) [reply testing_containers/ldap_sudo_environment/Dockerfile:4]: Thank you!

**Review by ddneilson — APPROVED** [review]

**Review by crowecawcaw — APPROVED** [review]

**Pull request merged**